docs/config.md

```diff
@@ -0,0 +1,51 @@
+# Configuration
+The philosophy we will use when exposing HCO configuration options is that only
+options that all operators consume can be exposed on the default HCO CR.
+Configuration options that are for specific component operators, will be hidden
+but accepted as valid configuration.
+
+Exposed variables are expressed in the HCO [CSV file](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/dfdd4ac492c1d91c130eec03af1e4f6b04d54c7e/deploy/converged/olm-catalog/kubevirt-hyperconverged/0.0.1/kubevirt-hyperconverged-operator.v0.0.1.clusterserviceversion.yaml#L18-L20)
+```yaml
```

Comment:
here's a thought. 

why not make the HCO spec an aggregate of all the operator CRs specs it encompasses? Something like this

```
spec: 
  kubeVirtSpec:
      imagePullPolicy: "IfNotPresent"
      imageTag: v0.15.0
      registry: "some-other-registry.io"
  cdiSpec:
      imagePullPolicy: "IfNotPresent"
.
.
.
```

In the code, the HCO would vendor the api package from each component which should already be represented in golang code as part of the operator SDK. Building the aggregated HCO spec would be a matter of mapping the structs vendored from those packages to the HCO api.  Updating the aggregated api to pick up new options exposed by a component level spec would only involve updating the vendored package for a specific component and rebuilding.

This would keep the HCO as dumb as possible. The only requirement on the component level operators is that they export their apis using golang, which everyone already gets for free with the Operator SDK.

Reply:
HCO should not care what the referenced operator does. And for example SSP does not have any golang code. We use operator sdk ansible.

Reply:
@davidvossel  We had similar discussion here: https://github.com/kubevirt/hyperconverged-cluster-operator/issues/13

Maybe my perception of HCO is wrong. I think HCO should not be dumb, it should be smart enough so I can afford to be dumb. i.e. there should be sane defaults and little to no configurables.

If we expose configuration of all underlying operators, what's the point of HCO.

Reply:
Btw, the hyperconverged operator should only expose features that are configurable by the user. Versions that are fixed for a release (compatibility) should definitely not be exposed. Or we could even support an override mechanism like you propose (hot fixes and debugging), but the default happy path should not depend on it.

Reply:
Not exposed as in "not documented" or "part of HCO manifest (environment variables?)"

Reply:
> If we expose configuration of all underlying operators, what's the point of HCO.
+1

>> Versions that are fixed for a release (compatibility) should definitely not be exposed. 
> Not exposed as in "not documented" or "part of HCO manifest (environment variables?)"
I should have said not required in the default use case. We might decide to allow overriding them for debugging or hot fix purposes.  But just for that, the admin should have no other reason to override almost anything. The ideal would be just CNV{version: 2.0} that gets translated to proper objects behind the scenes.

Reply:
> CNV{version: 2.0} 

I don't know what CNV is. I know what the HCO is though. It's an operator that lets me install everything related to the KubeVirt umbrella project through a single subscription. The versions and configurations used to install HCO (and all the underlying components making up HCO) are all up to me.

If someone was inclined to create a downstream product using the HCO, and they wanted to lock versions of all the components, then they'd have a specifically crafted HCO CR that meets their needs. 

> If we expose configuration of all underlying operators, what's the point of HCO.

HCO is a single entry point "meta" operator.  user subscribes to one thing (HCO) and gets everything (kubevirt, cdi, network addons, etc...)

> And for example SSP does not have any golang code. We use operator sdk ansible.

I'm unsure this will be a long term solution we can stick with, but that discussion can happen outside of this thread.

Reply:
So you are saying that HCO is not opinionated at all and just executes your manifest for you? That is just silly and means you do not need the HCO operator at all. Because you can post the same manifest manually instead.

HCO should be opinionated (in my opinion :). It should be a _preconfigured_ distribution of all the bits from kubevirt umbrella, including matching component versions and preselected config values for those.

Reply:
> So you are saying that HCO is not opinionated at all and just executes your manifest for you? 

yep

> That is just silly and means you do not need the HCO operator at all. Because you can post the same manifest manually instead.

HCO is a single marketplace entry, a single subscription that gets you everything. It also manages coordinating updates in an opinionated way.

> HCO should be opinionated (in my opinion :). It should be a preconfigured distribution of all the bits from kubevirt umbrella, including matching component versions and preselected config values for those.

We'd be shifting the opinionated configuration parts related to how the HCO CR is created outside the project to whatever consumes HCO. 

The HCO still maintains the ability to be opinionated on how the infrastructure is created and updated though. For example, we'd control the ordering each component level operator updates and handle coordinating avoiding any infrastructure related collisions that can occur there.

Reply:
> It also manages coordinating updates in an opinionated way.

How can it do that when it does not know which combinations are compatible or what the intermediate upgrade steps between versions are?

> We'd be shifting the opinionated configuration parts related to how the HCO CR is created outside the project to whatever consumes HCO.

So how does the operator help you if you still have to define everything yourself in the CR? That way you degrade it to manifest executor.

Btw: Did you notice line 21 where Ryan specifically talks about product versions?

Also, the opinionated consumer will pretty likely have to patch the source code logic too as the order of upgrades might be different based on other considerations they might have.

I really think we should make this an opinionated installation of KubeVirt pieces. With an optional way to customize it sure.. but it must work out of the box too.

Reply:
> I really think we should make this an opinionated installation of KubeVirt pieces. With an optional way to customize it sure.. but it must work out of the box too.

right, it has to work "out of the box". So, we ship HCO CR's that work "out of the box".

someone consuming HCO for a downstream project would ship their opinionated HCO CR.

Reply:
Just a note: Unlike kubevirt (correct me if I am wrong), manifests of network addons are a part of network addons operator, i.e. if you provide different version of images, it it likely that it won't work (e.g. requested image version wants to read X,Y, but RBAC manifests in operator allow only X).

Reply:
My goal for this PR is to establish a minimal list of "top level" parameters that the HCO will expose to the user and that each component operator should express on their CRs. 

That being said, I think the way I proposed we solve configuration has a flaw when consuming the HCO upstream.  @petrkotas pointed that out in [this discussion](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/22#discussion_r272352648).  Looking at @davidvossel's idea, I think it will work when consuming the HCO upstream, however I think this idea doesn't expose a "minimal list of top level parameters" - it's very granular.

In my opinion, we should only make visible the minimal configuration in the short term.  Long term, we can change to add more variables if needed.  But, I want to get agreement on the short term solution because it must be part of the long term solution otherwise we're breaking the API.

So the way I think this should work is in the HCO's first release, we provide a minimal set of top level parameters - version, imagePullPolicy, and containerRegistry.  `version` won't work upstream initially, which is ok because nothing will until all our operators are in operatorhub anyway.  Then after we cut the first release, we can evaluate how to solve this upstream with two options:
   1) separate downstream from upstream config.  We allow for more granular configuration by vendoring in component operator APIs (@davidvossel's idea), but those options are hidden by default.  In other words, the HCO's CR doesn't ship with them there by default, but we document their existence.  You can use all the config options both upstream or downstream, but upstream will only use the granular variables.
   2) we create a kubevirt umbrella release.  This requires building integration testing through the HCO to test upstream versions of all component operators that can be tied to `version` on the HCO CR.  (master will almost always be broken, but releases should be stable)

Reply:
> So the way I think this should work is in the HCO's first release, we provide a minimal set of top level parameters - version, imagePullPolicy, and containerRegistry.

I'm fine with this initially. For consistency, how would you feel about converging on ```imageTag, imageRegistry, and imagePullPolicy``` for names? That's how the KubeVirt CR exposes them right now.

Reply:
Guys you are imposing parameters on possibly 3rd party components. That will never work and is not acceptable to me. We might not have all the operators under our control (imagine if somebody decides to add cluster monitoring prometheus operator to the default deployment for example).

The only sustainable way is to collect all the manifests and CRs and use them in the provided form. Especially upstream where registry is not really important. The "product" maintainer can then update the compiled manifest in the operator of operators to make the next product version compilation.

This is pretty much the same as comps.xml for RPM based distributions (list of required RPMs to install). The RPMs are independent, have independent versioning and are responsible for the files. The distribution just puts and tests them together. Downstream distributions recompile the RPMs with different settings and so can the operator maintainer. You really do not need to care about downstream here too much (just make it easy to replace the defaults of the HCO, maintainers of the referenced operators will handle their side).

Reply:
Here's how I described component operator variables:
```
The component variable:
  - does __not__ have to be the same name as the HCO variable
  - can be expressed by multiple granular variables
```
Component operators don't need to have `version`, `imagePullPolicy`, and `containerRegistry` on their CRs.  The component variables are expressed on the component CR and are under control of the component operator developers. The HCO has a mapping from the "top level" to the component's variables.

The issues I created titled "Expose ContainerRegistry, ContainerTag, and PullPolicy on the operator's CR" muddied the waters a bit.  I changed s/Expose/Express/.  Kubevirt doesn't need to change anything because it already expresses those vars on the CR.